pkg/controller/servingcert/controller/secret_updating_controller.go

```diff
@@ -69,70 +70,105 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
 
 func (sc *serviceServingCertUpdateController) Sync(obj metav1.Object) error {
 	sharedSecret := obj.(*v1.Secret)
+	service := sc.getServiceForSecret(sharedSecret)
+	if service == nil {
+		return nil
+	}
 
-	regenerate, service := sc.requiresRegeneration(sharedSecret)
-	if !regenerate {
+	if !sc.isSecretValidForService(service, sharedSecret) {
 		return nil
 	}
 
 	// make a copy to avoid mutating cache state
 	secretCopy := sharedSecret.DeepCopy()
 
-	if err := toRequiredSecret(sc.dnsSuffix, sc.ca, service, secretCopy); err != nil {
+	update := false
+	servingCert, err := getServingCert(sc.dnsSuffix, sc.ca, service)
+	if err != nil {
 		return err
 	}
-
-	_, err := sc.secretClient.Secrets(secretCopy.Namespace).Update(secretCopy)
+	certBytes, keyBytes, err := servingCert.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+	if sc.requiresRegeneration(service, sharedSecret, sc.minTimeLeftForCert) {
+		if err := toRequiredSecret(servingCert, service, secretCopy); err != nil {
+			return err
+		}
+		update = true
+	} else {
+		data := map[string][]byte{
+			v1.TLSCertKey:       certBytes,
+			v1.TLSPrivateKeyKey: keyBytes,
+		}
+		// check that we only have the data we care about
+		// if new data does not match, regenerate it
+		if !bytes.Equal(secretCopy.Data[v1.TLSCertKey], data[v1.TLSCertKey]) ||
```

Comment:
I might be missing something but it seems since you're creating data new each time, it will contain a new cert+key and never match the existing secret, so this path will always lead to regeneration. 

In thinking through the different ways to modify the secret data and what our reaction to it in this else clause should be:
A. Someone adds in an extra data item. We want to get rid of this but keep the same tls.crt and tls.key, so for this we can have a standalone check for len(secretCopy.Data) > 2. If there is more than 2 items, then we make a new data map with the existing tls.crt and tls.key and update.
B. Someone edits tls.crt or tls.key data, replacing them with ascii art. Here we could x509 decode secretCopy data to verify that. If it's been munged then we update with the new cert data. If it's OK then we assume it's the serving cert. If someone replaced the cert data with a legitimate cert unless we had some kind of cache of the current certs issued by the serving CA (or we did an actual verify against the CA) we could not confirm that it belongs.

Reply:
makes sense, I'll update

Reply:
updated